chain/client/src/chunk_executor_actor.rs

```diff
@@ -0,0 +1,533 @@
+        }
+
+        if let Err(err) = self.try_apply_chunks(&block_hash, me.as_ref()) {
+            tracing::error!(target: "chunk_executor", ?err, ?block_hash, "failed to apply chunk for block hash");
+        };
+    }
+}
+
+impl ChunkExecutorActor {
+    #[instrument(target = "chunk_executor", level = "debug", skip_all, fields(%block_hash, ?me))]
+    fn try_apply_chunks(
+        &mut self,
+        block_hash: &CryptoHash,
+        me: Option<&AccountId>,
+    ) -> Result<(), Error> {
+        let epoch_id = self.epoch_manager.get_epoch_id(block_hash)?;
+        let block = self.chain_store.get_block(block_hash)?;
+        let header = block.header();
+        let prev_block_hash = header.prev_hash();
+        for shard_id in self.epoch_manager.shard_ids(&epoch_id)? {
+            let is_me = true;
+            if self
+                .shard_tracker
+                .cares_about_shard_this_or_next_epoch(me, block_hash, shard_id, is_me)
+                && !self.chain_store.incoming_receipts_exist(&prev_block_hash, shard_id)?
+            {
+                tracing::debug!(target: "chunk_executor", %block_hash, %prev_block_hash, "missing receipts to apply all tracked chunks for a block");
+                return Ok(());
+            }
+        }
+        self.apply_chunks(me, block, SandboxStatePatch::default())
+    }
+
+    fn get_incoming_receipts(
+        &self,
+        prev_block_hash: &CryptoHash,
+        shard_id: ShardId,
+    ) -> Result<Arc<Vec<ReceiptProof>>, Error> {
+        self.chain_store.get_incoming_receipts(prev_block_hash, shard_id)
+    }
+
+    // Logic here is based on Chain::apply_chunk_preprocessing
+    fn apply_chunks(
+        &mut self,
+        me: Option<&AccountId>,
+        block: Block,
+        mut state_patch: SandboxStatePatch,
+    ) -> Result<(), Error> {
+        let block_hash = block.hash();
+        let header = block.header();
+        let prev_hash = header.prev_hash();
+        let prev_block = self.chain_store.get_block(prev_hash)?;
+
+        let prev_chunk_headers = self.epoch_manager.get_prev_chunk_headers(&prev_block)?;
+
+        let epoch_id = block.header().epoch_id();
+        let shard_layout = self.epoch_manager.get_shard_layout(&epoch_id)?;
+
+        let chunk_headers = &block.chunks();
+        let mut jobs = Vec::new();
+        // TODO(spice-resharding): Make sure shard logic is correct with resharding.
+        for (shard_index, _prev_chunk_header) in prev_chunk_headers.iter().enumerate() {
+            // XXX: This is a bit questionable -- sandbox state patching works
+            // only for a single shard. This so far has been enough.
+            let state_patch = state_patch.take();
+            let shard_id = shard_layout.get_shard_id(shard_index)?;
+
+            let chunk_header =
+                chunk_headers.get(shard_index).ok_or(Error::InvalidShardId(shard_id))?;
+
+            let is_new_chunk = chunk_header.is_new_chunk(block.header().height());
```

Comment:
Should this always be true? We **do** want to apply the chunk even if the chunk doesn't exist - different from the old behavior. Or are we not changing that at the moment?

Reply:
Good catch. I don't remember why in my previous versions this didn't work with is new chunk true - likely something else was wrong. Changed the code so that is_new_chunk is always true as part of [`ff7fe8a` (#13637)](https://github.com/near/nearcore/pull/13637/commits/ff7fe8a54d011c45b5bced06ae2b78377001efe7)